python/cuspatial/cuspatial/geometry/geoseries.py

```diff
@@ -0,0 +1,457 @@
+        while mp_index >= 0:
+            if self.source.types[mp_index] == "mpoly":
+                preceding_mp_count = preceding_mp_count + 1
+            elif (
+                self.source.types[mp_index] == "poly"
+                and preceding_mp_count == 0
+            ):
+                preceding_poly_count = preceding_poly_count + 1
+            mp_index = mp_index - 1
+        preceding_multis = preceding_mp_count
+        multi_index = (
+            self.source._polygons.mpolys[preceding_multis * 2 - 1]
+            if preceding_multis > 0
+            else 0
+        )
+        preceding_polys = preceding_poly_count
+        ring_start = self.source._polygons.polys[multi_index + preceding_polys]
+        ring_end = self.source._polygons.polys[
+            multi_index + preceding_polys + 1
+        ]
+        rings = self.source._polygons.rings
+        exterior_slice = slice(rings[ring_start], rings[ring_start + 1])
+        exterior = self.source._polygons.xy[exterior_slice]
+        interiors = []
+        for interior in range(ring_end - (ring_start + 1)) + (ring_start + 1):
+            interior_slice = slice(rings[interior], rings[interior + 1])
+            interiors.append(
+                self.source._polygons.xy[interior_slice]
+                .to_array()
+                .reshape(int((rings[interior + 1] - rings[interior]) / 2), 2)
+            )
+        return Polygon(
+            exterior.to_array().reshape(2 * (ring_start - ring_end), 2),
+            interiors,
+        )
+
+
+class cuMultiPolygon(cuGeometry):
+    def to_shapely(self):
+        item_type = self.source.types[self.index]
+        index = (
+            pd.Series(self.source.types[0 : self.index]) == item_type
+        ).sum()
+        poly_indices = slice(
+            self.source.polygons.mpolys[index * 2],
+            self.source.polygons.mpolys[index * 2 + 1],
+        )
+        polys = []
+        for i in range(poly_indices.start, poly_indices.stop):
+            ring_start = self.source._polygons.polys[i]
+            ring_end = self.source._polygons.polys[i + 1]
+            rings = self.source._polygons.rings
+            exterior_slice = slice(rings[ring_start], rings[ring_start + 1])
+            exterior = self.source._polygons.xy[exterior_slice]
+            interiors = []
+            for interior in range(ring_start + 1, ring_end):
+                interior_slice = slice(rings[interior], rings[interior + 1])
+                interiors.append(
+                    self.source._polygons.xy[interior_slice]
+                    .to_array()
+                    .reshape(
+                        int((rings[interior + 1] - rings[interior]) / 2), 2
+                    )
+                )
+            polys.append(
+                Polygon(
+                    exterior.to_array().reshape(
+                        2 * (ring_start - ring_end), 2
+                    ),
+                    interiors,
+                )
+            )
```

Comment:
super nit: Could turn this into a List comprehension

Reply:
:heavy_check_mark: